### PR TITLE
Add more specs for `YAML::Any#[]` and `#[]?`

### PR DESCRIPTION
**Pull request by straight-shoota**
This patch just adds some specs for `Any#[]` and `Any#[]?` which test with all YAML types as keys/indices.

Note: This just describes the current behaviour which has some shady edges. I'd like to merge this first and then work on several individual enhancements.

### CODE REVIEW

**Comment by wontruefree** [comment]
The purpose of these specs are not super obvious to me.  it seems like each test case is testing multiple edge cases and the naming is fairly generic.  I think breaking the test cases up might help with legibility.

File: spec/std/yaml/any_spec.cr [hunk 1]
```diff
@@ -2,6 +2,92 @@ require "spec"
...
+
+    equivalent_keys.each do |k|
+      any[k].raw.should eq("bar")
+      # FIXME: Can't do `YAML::Any.new` with arbitrary number types (#11645)
+      if k.is_a?(YAML::Any::Type)
+        any[YAML::Any.new(k)].raw.should eq("bar")
+      end
+    end
+
+    unless key.nil?
```
[the remark this answers is not in the document]
Comment (straight-shoota) [reply spec/std/yaml/any_spec.cr:20]: The following specs test for *other* keys that might match the type, but not value of `key`. The `Nil` type has only a single value, so this doesn't apply and we must make an exception. There's no way to test a different value for that type.

**Review by beta-ziliani — APPROVED** [review]
